akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -175,27 +188,84 @@ private[akka] class Shard(typeName: String,
+      case t ⇒ LeaseAcquireResult(acquired = false, Some(t))
+    }).to(self)
+  }
 
   def processChange[E <: StateChange](event: E)(handler: E => Unit): Unit =
     handler(event)
 
   def receive = receiveCommand
 
+  // Don't send back ShardInitialized so that messages are buffered in the ShardRegion
+  // while awaiting the lease
+  def awaitingLease: Receive = {
+    case LeaseAcquireResult(true, _) ⇒
+      log.info("Acquired lease. Ready to accept commands.")
+      context.parent ! ShardInitialized(shardId)
+      context.become(receiveCommand)
+    case LeaseAcquireResult(false, None) ⇒
+      log.error("Failed to get lease for shard type [{}] id [{}]. Retry in {}", typeName, shardId, leaseRetryInterval)
+      timers.startSingleTimer(LeaseRetryTimer, LeaseRetry, leaseRetryInterval)
+    case LeaseAcquireResult(false, Some(t)) ⇒
+      log.error(t,
+                "Failed to get lease for shard type [{}] id [{}]. Retry in {}",
+                typeName,
+                shardId,
+                leaseRetryInterval)
+      timers.startSingleTimer(LeaseRetryTimer, LeaseRetry, leaseRetryInterval)
+    case LeaseRetry ⇒
+      tryGetLease(lease.get)
+    case ll: LeaseLost ⇒
+      receiveLeaseLost(ll)
+  }
+
   def receiveCommand: Receive = {
-    case Terminated(ref)                         => receiveTerminated(ref)
-    case msg: CoordinatorMessage                 => receiveCoordinatorMessage(msg)
-    case msg: ShardCommand                       => receiveShardCommand(msg)
-    case msg: ShardRegion.StartEntity            => receiveStartEntity(msg)
-    case msg: ShardRegion.StartEntityAck         => receiveStartEntityAck(msg)
-    case msg: ShardRegionCommand                 => receiveShardRegionCommand(msg)
-    case msg: ShardQuery                         => receiveShardQuery(msg)
-    case PassivateIdleTick                       => passivateIdleEntities()
-    case msg if extractEntityId.isDefinedAt(msg) => deliverMessage(msg, sender())
+    case Terminated(ref) ⇒ receiveTerminated(ref)
+    case msg: CoordinatorMessage ⇒ receiveCoordinatorMessage(msg)
+    case msg: ShardCommand ⇒ receiveShardCommand(msg)
+    case msg: ShardRegion.StartEntity ⇒ receiveStartEntity(msg)
+    case msg: ShardRegion.StartEntityAck ⇒ receiveStartEntityAck(msg)
+    case msg: ShardRegionCommand ⇒ receiveShardRegionCommand(msg)
+    case msg: ShardQuery ⇒ receiveShardQuery(msg)
+    case PassivateIdleTick ⇒ passivateIdleEntities()
+    case msg: LeaseLost ⇒ receiveLeaseLost(msg)
+    case msg if extractEntityId.isDefinedAt(msg) ⇒ deliverMessage(msg, sender())
   }
 
+  def receiveLeaseLost(msg: LeaseLost): Unit = {
+    // The shard region will re-create this when it receives a message for this shard
+    log.error("Shard type [{}] id [{}] lease lost. Reason: {}", typeName, shardId, msg.reason)
+    // should we try and stop entities with the stop message or just stop ASAP?
```

Comment:
not sure but I think a quick stop might be better to avoid processing more messages in the entities, although it means dropping messages in flight

Reply:
Will leave it as is then